### PR TITLE
Implement tagged unions for named fields

### PR DESCRIPTION
**Pull request by LostKobrakai**
This is mostly a small change, but I needed to add another parameter to `decode` / `do_decode`, which resulted in quite a lot of LOC changed. I'm not sure if this is the best way to bring the option to the code decoding unions though.

### CODE REVIEW

File: lib/avro_ex/decode.ex [hunk 1]
```diff
@@ -172,25 +177,31 @@ defmodule AvroEx.Decode do
...
+  defp do_decode(%Union{possibilities: possibilities}, %Context{} = context, data, opts)
        when is_binary(data) do
-    {index, rest} = do_decode(%Primitive{type: :long}, context, data)
+    {index, rest} = do_decode(%Primitive{type: :long}, context, data, opts)
     schema = :lists.nth(index + 1, possibilities)
 
-    do_decode(schema, context, rest)
+    {decoded_item, rest} = do_decode(schema, context, rest, opts)
+
+    if Keyword.get(opts, :tagged_unions, false) && Map.has_key?(schema, :name) do
```
Comment (davydog187) [inline lib/avro_ex/decode.ex:191]: Nitpick

```suggestion
    if Keyword.get(opts, :tagged_unions, false) and Map.has_key?(schema, :name) do
```

File: lib/avro_ex.ex [hunk 2]
```diff
@@ -155,11 +155,11 @@ defmodule AvroEx do
       {:ok, true}
 
   """
-  @spec decode(Schema.t(), encoded_avro) ::
+  @spec decode(Schema.t(), encoded_avro, keyword()) ::
```
Comment (davydog187) [inline lib/avro_ex.ex:158]: We should update the respective documentation

File: test/encode_test.exs [hunk 3]
```diff
@@ -285,6 +285,32 @@ defmodule AvroEx.Encode.Test do
...
+
+      json_schema = ~s([#{record_json_factory.("a")}, #{record_json_factory.("b")}])
+
+      {:ok, schema} = AvroEx.decode_schema(json_schema)
+      {:ok, int_schema} = AvroEx.decode_schema(~S("int"))
+      {:ok, record_schema} = AvroEx.decode_schema(record_json_factory.("b"))
+
+      {:ok, index} = @test_module.encode(int_schema, 1)
+      {:ok, encoded_record} = @test_module.encode(record_schema, %{"value" => "hello"})
+      {:ok, encoded_union} = @test_module.encode(schema, {"b", %{"value" => "hello"}})
```
Comment (davydog187) [inline test/encode_test.exs:309]: can you add a test case for the unhappy path?

Comment (LostKobrakai) [reply test/encode_test.exs:309]: This has been added

File: lib/avro_ex.ex (hunk 2 above)
Comment (LostKobrakai) [reply lib/avro_ex.ex:172]: I've added explanations to the feature. Wasn't really sure if there's a good way to have examples about those, given how big the schema alone would be.

**Review by davydog187 — APPROVED** [review]
This looks great, thank you @LostKobrakai!

**Pull request merged**